### PR TITLE
Implemented concat_uniform_length

### PR DESCRIPTION
**Pull request by 702fbtngus**
Until now, the instruction `array.new_data` was implemented manually, in file spectec/src/il2al/manual.ml. Now when converting it so that it is implemented automatically from DSL, usage of the function `inverse_of_concat` became a problem. It always splits the input list into chunks of length 2, but `array.new_data` needed the function to split it into chunks of some other length, which is `$($zsize(zt)/8)`. So I implemented a new function similar to current `concat_`, named `concat_uniform_length`, with an additional parameter which indicates the length of a chunk it should split.
Using `concat_uniform_length` in definition of `array.new_data` works well.

Here are some following thoughts:
* Is the name `concat_uniform_length` appropriate? Maybe too long, since there is now a new function named `inverse_of_concat_uniform_length(_helper)` in file spectec/src/backend-interpreter/numerics.ml.
* Or will it be better to just make a function called, for example, `split_by_length`, and use it instead of `inverse_of_concat`, like below?
```
;; FROM
-- if $concat_uniform_length(byte, $zbytes_(zt, c)^n, $($zsize(zt)/8)) = $data(z, y).BYTES[i : n * $zsize(zt)/8]
;; TO
-- if $zbytes_(zt, c)^n = $split_by_length(byte, $($zsize(zt)/8)), $data(z, y).BYTES[i : n * $zsize(zt)/8])
```
* I've tried to replace every `concat_` into `concat_uniform_length` with chunk length 2, where `inverse_of_concat` is executed instead of `concat` in interpreter. There were two of those cases, `vextunop` and `vextbinop`. However replacing them caused an error as below. Why is it happening, and how can I deal with it?
```
def $vextunop(Jnn_1 X N_1, Jnn_2 X N_2, EXTADD_PAIRWISE sx, c_1) = c
  -- var cj_1 : iN($lsize(Jnn_2))
  -- var cj_2 : iN($lsize(Jnn_2))
  -- if ci* = $lanes_(Jnn_1 X N_1, c_1)
  -- if $concat_uniform_length(iN($lsize(Jnn_2)), (cj_1 cj_2)*, 2) = $ext($lsize(Jnn_1), $lsize(Jnn_2), sx, ci)*
  -- if c = $invlanes_(Jnn_2 X N_2, $iadd($lsize(Jnn_2), cj_1, cj_2)*)
```
```
spec/wasm-3.0/3-numerics.watsup:438.1-443.71: type error: definition contains indeterminate variable(s) `cj_1`
```

### CODE REVIEW

**Comment by rossberg** [comment]
Okay, if I understand correctly, the problem is that the concat function is not an invertible bijection, since many different arguments can produce the same result. And adding a size parameter would make it one. That makes sense.

I'd indeed suggest replacing all uses of concat, i.e., change its definition to
```
def $concat__(syntax X, nat, (X*)*) : X*  hint(show (++) %3)
def $concat__(syntax X, n, eps) = eps
def $concat__(syntax X, n, (w^n) (w'^n)*) = w^n $concat__(X, n, (w'^n)*)
```
Note that this is actually restricting the parts to have length n, otherwise the inversion would still not be correct semantically.

To answer the last question: the cj's do not occur in anything considered a "binding" position. Those are only in "patterns" left or right of an equation, or as last argument to a function call (to allow for inversion). This is primarily a pragmatic sanity check. If you make the size argument the second parameter, it's accepted.

**Comment by 702fbtngus** [comment]
I've tried replacing all uses of concat so that it restricts the parts to have length n, however it fails when replacing a usage in function instantiate.
```
  -- if instr_E* = $concat_(instr, $lenelem_(elem*), $runelem_(i_E, elem*[i_E])^(i_E<|elem*|))
  -- if instr_D* = $concat_(instr, $lendata_(data*), $rundata_(i_D, data*[i_D])^(i_D<|data*|))
```
Here, it should concatenate the parts of the result of runelem and rundata. (lenelem and lendata are made to handle the cases where |elem*| or |data*| is 0)
However, it seems to make error since the lengths of parts of the result are not identical.

So I think the length parameter for concat is not needed in case it is used directly (I mean, not for its inversion). I could think of two solutions:
* Just separating concat into two functions, where one is directly used without length parameter and the other is used with length parameter for actually splitting the argument (like before)
* Make length parameter optional for concat and remove the length restriction in direct case, as below:
```
def $concat_(syntax X, nat?, (X*)*) : X*  hint(show (++) %3)
def $concat_(syntax X, n?, eps) = eps
def $concat_(syntax X, n, (w^n) (w'^n)*) = w^n $concat_(X, n, (w'^n)*)
def $concat_(syntax X, eps, (w*) (w'*)*) = w^n $concat_(X, eps, (w'*)*)
```
```
  -- if instr_E* = $concat_(instr, eps, $runelem_(i_E, elem*[i_E])^(i_E<|elem*|))
  -- if instr_D* = $concat_(instr, eps, $rundata_(i_D, data*[i_D])^(i_D<|data*|))
```

**Comment by 702fbtngus** [comment]
Great! I've modified the code and pushed on the branch.

File: spectec/spec/wasm-3.0/0-aux.watsup [hunk 1]
```diff
@@ -41,6 +41,10 @@ def $concat_(syntax X, (X*)*) : X*  hint(show (++) %2)
 def $concat_(syntax X, eps) = eps
 def $concat_(syntax X, (w*) (w'*)*) = w* $concat_(X, (w'*)*)
 
+def $concatn(syntax X, (X*)*, nat) : X*  hint(show (++) %2)
+def $concatn(syntax X, eps, n) = eps
+def $concatn(syntax X, (w*) (w'*)*, n) = w* $concatn(X, (w'*)*, n)
```
Comment (rossberg) [inline spectec/spec/wasm-3.0/0-aux.watsup:46]:
```suggestion
def $concatn_(syntax X, (X*)*, nat) : X*  hint(show (++) %2)
def $concatn_(syntax X, eps, n) = eps
def $concatn_(syntax X, (w^n) (w'^n)*, n) = w* $concatn_(X, (w'^n)*, n)
```

**Review by rossberg — APPROVED** [review]
LGTM modulo comment

File: spectec/spec/wasm-3.0/0-aux.watsup (hunk 1 above)
Comment (702fbtngus) [reply spectec/spec/wasm-3.0/0-aux.watsup]: Great

**Pull request merged**